sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala

```diff
@@ -131,4 +134,23 @@ class FileScanRDD(
   }
 
   override protected def getPartitions: Array[RDDPartition] = filePartitions.toArray
+
+  override protected def getPreferredLocations(split: RDDPartition): Seq[String] = {
+    val files = split.asInstanceOf[FilePartition].files
+
+    // Computes total number of bytes can be retrieved from each host.
+    val hostToNumBytes = mutable.HashMap.empty[String, Long]
+    files.foreach { file =>
+      file.locations.filter(_ != "localhost").foreach { host =>
```

Comment:
We should filter them out. For a partition that doesn't have any preferred locations, it can be bundled with any other tasks and scheduled to any executor. But once it's marked with "localhost", delayed scheduling may be triggered because they have different host name as other tasks. Further more, "localhost" isn't a valid location for the `DAGScheduler` when deciding which executors to run the tasks.